crates/ide_assists/src/handlers/merge_match_arms.rs

```diff
@@ -48,10 +48,14 @@ pub(crate) fn merge_match_arms(acc: &mut Assists, ctx: &AssistContext) -> Option
             if arm.guard().is_some() {
                 return false;
             }
-            match arm.expr() {
-                Some(expr) => expr.syntax().text() == current_expr.syntax().text(),
-                None => false,
+            if let Some(expr) = arm.expr() {
+                if let Some(expr_ty) = ctx.sema.type_of_expr(&expr) {
+                    if let Some(current_expr_ty) = ctx.sema.type_of_expr(&current_expr) {
+                        return expr.syntax().text() == current_expr.syntax().text() && expr_ty == current_expr_ty
+                    }
+                }
             }
+            false
         })
```

Comment:
The original issue talks about the types of identifiers bound in the patterns, not the types of the expressions.

That is it merges the arms in this case
```rs
match Result::<i32, f32>::Ok(0) {
    Ok(x) => (),
    Err(x) => (),
}
```
but it should not, as the `x` is of different type in each.

Checking the type of the expressions isnt really needed as it would be an error for the types to differ anyways.

(Also aplogies for keeping this unreviewed for so long)